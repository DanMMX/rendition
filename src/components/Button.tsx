import * as React from 'react';
import styled, { StyledFunction, withTheme } from 'styled-components';
import hoc from '../hoc';
import { normal, bold, darken, getColor, getColoringType, px } from '../utils';

interface ButtonProps extends DefaultProps, Coloring, Sizing {
	square?: boolean;
	disabled?: boolean;
	outline?: boolean;
	plaintext?: boolean;
	underline?: boolean;
}

const squareWidth = (val: number): number => val / 9 * 10;

const minWidth = (props: ButtonProps) => {
	if (props.w == null && !props.square) {
		return 'auto';
	}
	if (props.square) {
		return props.emphasized
			? squareWidth(props.theme.space[5])
			: squareWidth(props.theme.space[4]);
	}

	return props.w;
};

const horizontalPadding = (props: ButtonProps) => {
	if (props.w == null && !props.square) {
		return props.emphasized ? 50 : 20;
	}
	if (props.square || props.w) {
		return 0;
	}

	return props.emphasized ? 50 : 20;
};

const styleableButton: StyledFunction<ButtonProps> = styled.button;

const Button = styleableButton`
	padding-top: 1px;
	padding-left: ${props => px(horizontalPadding(props))};
	padding-right: ${props => px(horizontalPadding(props))};
	font-family: inherit;
	display: inline-block;
	font-weight: ${props => bold(props)};
	border-radius: ${props => px(props.theme.radius)};
	appearance: none;
	text-decoration: none;
	border: 0;
	margin: 0;
	min-width: ${props => px(minWidth(props))};
	vertical-align: middle;
	font-size: inherit;
	line-height: 1.1;
	text-align: center;
	cursor: pointer;
	background: ${props => props.bg || getColor(props, 'bg', 'main')};
	color: ${props => props.color || '#fff'};
	height: ${(props: ButtonProps) =>
		px(props.emphasized ? props.theme.space[5] : props.theme.space[4])};

	&:hover,
	&:focus,
	&:active {
		color: #fff;
		background-color: ${props =>
			getColor(props, 'bg', 'dark') || darken(props.color as string)};
	}

	&:disabled {
		opacity: 0.65;
		cursor: not-allowed;
	}
`;

const Outline = Button.extend`
	color: ${props =>
		getColor(props, 'color', 'main') || props.theme.colors.text.main};
	background: ${props => props.color || 'none'};
	border: 1px solid;
`;

const Plaintext = Button.extend`
	padding-left: 0;
	padding-right: 0;
	height: auto;
	border: 0;
	border-radius: 0;
	color: ${props =>
		getColor(props, 'color', 'main') || props.theme.colors.text.main};
	background: ${props => props.color || 'none'};
	font-weight: ${props => normal(props)};

	&:hover,
	&:focus,
	&:active {
		background: none;
		color: ${props =>
			getColor(props, 'color', 'dark') || props.theme.colors.text.main};
	}
`

const Underline = Plaintext.extend`
	padding-bottom: 2px;
	border-bottom: 1px solid;
	font-weight: ${props => bold(props)};

	&:hover,
	&:focus,
	&:active {
		color: ${props =>
			getColor(props, 'color', 'main') || props.theme.colors.text.main};
		box-shadow: 0px -1px 0 0px inset;
	}
`;

export default withTheme(
	hoc(({ outline, underline, plaintext, ...props }: ButtonProps) => {
		if (plaintext) {
			return <Plaintext {...props} />;
		} else if (outline) {
			return <Outline {...props} />;
		} else if (underline) {
			return <Underline {...props} />;
		} else if (!getColoringType(props) && !props.color && !props.bg) {
			// outline tertiary is our default btn
			return <Outline {...props} tertiary />;
		} else {
			return <Button {...props} />;
		}
	}),
);
