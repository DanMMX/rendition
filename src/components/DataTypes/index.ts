import * as booleanType from './boolean';
import * as dateTimeType from './date-time';
import * as enumType from './enum';
import * as numberType from './number';
import * as objectType from './object';
import * as stringType from './string';

export default {
	boolean: booleanType,
	date_time: dateTimeType,
	enum: enumType,
	number: numberType,
	object: objectType,
	string: stringType,
};
