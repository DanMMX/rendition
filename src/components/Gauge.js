import * as React from 'react'
const Pie = require('paths-js/pie')
import styled, { withTheme } from 'styled-components'
const map = require('lodash/map')
import { compose } from 'recompose'
import { px } from '../utils'

import asRendition from '../asRendition'
import theme from '../theme'

const { font } = theme

const StatusGaugeWrapper = styled.div`
  display: block;
  color: ${({ theme }) => theme.colors.text.main};

  /* fix edge hover missplacing bug
    on svg text inside anchors
  */
  &,
  &:hover {
    text-decoration: none;

    text {
      text-decoration: none !important;
    }

    .gauge__graph__total-label {
      transform: translateX(0);
    }
  }

  .gauge__graph {
    display: block;
    width: 100%;
    max-width: 180px;
    margin: 0 auto 10px;
  }

  .gauge__graph__total-label {
    font-family: ${font};
    font-size: 5px;
    fill: #b8b8b8;
    text-anchor: middle;
  }

  .gauge__legend {
    list-style: none;
    padding: 5px;
    margin: 0 auto;
    background: white;
    border-radius: 5px;
    max-width: 240px;
    border: 1px solid #ececec;
    overflow: hidden;
  }

  .gauge__legend-item {
    display: flex;
    padding: 3px 0 3px 2px;
    line-height: 9px;
    width: 50%;
    float: left;
  }

  .gauge__legend-color {
    display: inline-block;
    flex: 0 0 auto;
    width: 10px;
    height: 10px;
    margin-right: 4px;
  }

  .gauge__legend-name {
    font-size: 11px;
  }
`

const GaugeTotalText = styled.text`
  font-family: ${font};
  font-size: ${props => px(props.count > 999 ? 20 : 26)};
  fill: #5e5e5e;
  letter-spacing: -1px;
  text-anchor: middle;
  transform: translateX(0);
`

const GaugeTotal = ({ count, ...props }) => (
  <GaugeTotalText count={count} {...props}>
    {count}
  </GaugeTotalText>
)

/**
 * data: [{
 *   key: string
 *   value: number
 *   color: string
 *   name: string
 * }]
 */
const StatusGauge = ({
  placeholderColor = theme.colors.gray.light,
  legend,
  data,
  title,
  ...props
}) => {
  const count = data.reduce((carry, item) => {
    return carry + item.value
  }, 0)

  const graph = Pie({
    accessor: item => item.value,
    center: [0, 0],
    data: data,
    R: 46,
    r: 25
  })

  const gauge = graph.curves
  return (
    <StatusGaugeWrapper {...props}>
      <svg
        className='gauge__graph'
        viewBox='0 0 100 100'
        xmlns='http://www.w3.org/2000/svg'
      >
        <circle
          fill=' white'
          stroke='#dadada'
          strokeWidth='5'
          cx='50'
          cy='50'
          r='46'
        />

        <g>
          <circle fill={placeholderColor} cx='50' cy='50' r='46' />
          <circle fill='white' cx='50' cy='50' r='25' />
        </g>

        <g transform='translate(50, 50)'>
          {map(gauge, segment => {
            if (!count) {
              return
            }
            if (segment.item.value === count) {
              return (
                <g key={segment.item.key || segment.item.name}>
                  <path
                    transform='translate(-100, -100)'
                    fill={segment.item.color}
                    d='M100 100m-46,0a46,46,0 1,0 92,0a 46,46 0 1,0 -92,0zM100 100m-25,0a25,25,0 0,1 50,0a 25,25 0 0,1 -50,0z'
                  />
                </g>
              )
            }
            return (
              <g key={segment.item.key || segment.item.name}>
                <path
                  fill={segment.item.color}
                  d={segment.sector.path.print()}
                />
              </g>
            )
          })}
        </g>
        <text className='gauge__graph__total-label' x='50' y='41'>
          {title ? ` ${title.toUpperCase()}` : 'TOTAL'}
        </text>
        <GaugeTotal x='50' y='63' count={count} />
      </svg>

      <ul className='gauge__legend'>
        {map(data, item => {
          return (
            <li key={item.key || item.name} className={`gauge__legend-item`}>
              <span
                className={`gauge__legend-color`}
                style={{ backgroundColor: item.color }}
              >
                &nbsp;
              </span>
              <span className='gauge__legend-name'>{item.name}</span>
            </li>
          )
        })}
      </ul>
    </StatusGaugeWrapper>
  )
}

export default compose(withTheme, asRendition)(StatusGauge)
