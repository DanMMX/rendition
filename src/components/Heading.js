import styled from 'styled-components'
import { color, fontSize, space } from 'styled-system'
import asRendition from '../asRendition'
import { monospace } from '../utils'
import { align, bold, caps } from './Txt'

const Heading = styled.h3`
  ${align}
  ${color}
  ${fontSize}
  ${monospace};
  ${space}

  ${caps}
  ${bold}
`

Heading.displayName = 'Heading'

Heading.defaultProps = {
  fontSize: 4,
  m: 0
}

const Base = asRendition(Heading)

Base.h1 = asRendition(Heading.withComponent('h1'))
Base.h1.defaultProps = {
  fontSize: 6,
  m: 0
}

Base.h2 = asRendition(Heading.withComponent('h2'))
Base.h2.defaultProps = {
  fontSize: 5,
  m: 0
}

Base.h3 = asRendition(Heading.withComponent('h3'))
Base.h3.defaultProps = {
  fontSize: 4,
  m: 0
}

Base.h4 = asRendition(Heading.withComponent('h4'))
Base.h4.defaultProps = {
  fontSize: 3,
  m: 0
}

Base.h5 = asRendition(Heading.withComponent('h5'))
Base.h5.defaultProps = {
  fontSize: 2,
  m: 0
}

Base.h6 = asRendition(Heading.withComponent('h6'))
Base.h6.defaultProps = {
  fontSize: 1,
  m: 0
}

export default Base
