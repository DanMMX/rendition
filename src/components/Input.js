import styled, { withTheme } from 'styled-components'
import asRendition from '../asRendition'
import { monospace, px } from '../utils'
import { radius } from '../theme'
import { compose } from 'recompose'

const getColor = props => {
  if (props.invalid) {
    return props.theme.colors.danger.main
  }
  if (props.valid) {
    return props.theme.colors.success.main
  }
  if (props.disabled) {
    return props.theme.colors.gray.light
  }
  return props.theme.colors.gray.main
}

const Base = styled.input`
  border-radius: ${px(radius)};
  height: ${props =>
    px(props.emphasized ? props.theme.space[5] : props.theme.space[4])};
  font-size: inherit;
  border: 1px solid ${props => getColor(props)};
  padding: 0px 16px;

  &[disabled] {
    background-color: ${props => props.theme.colors.gray.light};
  }
  &:focus {
    border-color: ${props => props.theme.colors.secondary.main};
  }
  &:hover {
    box-shadow: 0 0 4px 1px rgba(0, 0, 0, 0.1);
  }
  &::placeholder {
    color: ${props => props.theme.colors.gray.main};
  }

  &[type="checkbox"] {
    height: auto;
    font-size: ${props => px(props.theme.fontSizes[1])};
  }

  ${monospace};
`

export default compose(withTheme, asRendition)(Base)
