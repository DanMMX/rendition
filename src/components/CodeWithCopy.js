import * as React from 'react'
import styled, { withTheme } from 'styled-components'
import * as copyToClipboard from 'copy-to-clipboard'
import { compose } from 'recompose'
import * as FaClipboard from 'react-icons/lib/fa/clipboard'
import Button from './Button'
import hoc from '../hoc'
import { stopEvent } from '../utils'

const Wrapper = styled.span`
  white-space: initial;

  code {
    font-family: ${props => props.theme.monospace};
    padding: 2px 4px;
    font-size: 90%;
    color: #c7254e;
    background-color: #f9f2f4;
    border-radius: 2px;
    white-space: normal;
    word-wrap: break-word;
    font-size: 1em;
    margin-right: 3px;
  }

  .code-with-copy__copy {
    .text-selection--none;
    cursor: pointer;
  }
`

const copyIt = (copy, text) => {
  const copyText = (copy || text || '').trim()

  copyToClipboard(copyText)
}

const Base = ({ copy, text, color, ...props }) => {
  return (
    <Wrapper {...props} className='code-with-copy'>
      <code title={copy}>{text.trim()}</code>

      <span onClick={stopEvent}>
        <Button
          plaintext
          tooltip={{ text: 'Copied!', trigger: 'click' }}
          onClick={() => copyIt(copy, text)}
          className='code-with-copy__copy'
        >
          <FaClipboard color={color} />
        </Button>
      </span>
    </Wrapper>
  )
}

export default compose(withTheme, hoc)(Base)
