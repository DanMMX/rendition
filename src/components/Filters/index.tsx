import cloneDeep = require('lodash/cloneDeep');
import findIndex = require('lodash/findIndex');
import map = require('lodash/map');
import reject = require('lodash/reject');
import * as React from 'react';
import { FaFilter, FaSearch } from 'react-icons/lib/fa';
import { FiltersProps, FiltersView } from 'rendition';
import styled from 'styled-components';
import Theme from '../../theme';
import * as utils from '../../utils';
import Button from '../Button';
import { getDataModel } from '../DataTypes';
import DeleteBtn from '../DeleteButton';
import { Box, Flex } from '../Grid';
import Modal from '../Modal';
import Select from '../Select';
import Txt from '../Txt';
import * as SchemaSieve from './SchemaSieve';
import ViewsMenu from './ViewsMenu';

import { JSONSchema6 } from 'json-schema';

import Summary from './Summary';

interface FilterInputProps {
	schema: JSONSchema6;
	value: any;
	operator: string;
	onUpdate: (value: any) => void;
}

const FilterInput = (props: FilterInputProps) => {
	const model = getDataModel(props.schema);

	if (!model) {
		return null;
	}

	return (
		<model.Edit
			schema={props.schema}
			value={props.value}
			operator={props.operator}
			onUpdate={props.onUpdate}
			slim
		/>
	);
};

const RelativeBox = styled(Box)`
	position: relative;
`;

const ExtraRuleDeleteBtn = styled(DeleteBtn)`
	position: absolute;
	bottom: 7px;
	right: -35px;
`;

const SimpleSearchBox = styled.div`
	position: relative;
	width: 500px;
	border-bottom: 2px solid ${Theme.colors.gray.main};
	padding-left: 20px;
	padding-top: 3px;
	margin-left: 30px;
	margin-right: 30px;

	.search-icon {
		position: absolute;
		top: 7px;
		left: 0;
		color: ${Theme.colors.gray.main};
	}

	input {
		background: transparent;
		box-shadow: none;
		border: none;
		width: 100%;
		font-size: inherit;
		padding: 5px;
		::placeholder {
			font-style: italic;
			color: ${Theme.colors.gray.main};
		}
	}
`;

const FilterWrapper = styled(Box)`
	position: relative;
`;

interface EditModel {
	field: string;
	operator: string;
	value: string | number | { [k: string]: string };
}

interface FiltersState {
	showModal: boolean;
	edit: EditModel[];
	editingFilter: string | null;
	searchString: string;
	filters: JSONSchema6[];
	views: FiltersView[];
}

class Filters extends React.Component<FiltersProps, FiltersState> {
	constructor(props: FiltersProps) {
		super(props);
		const { filters, views } = this.props;

		this.state = {
			showModal: false,
			searchString: '',
			edit: [this.getCleanEditModel()],
			editingFilter: null,
			filters: filters || [],
			views: views || [],
		};

		// Clean exsting rules on load
	}

	getCleanEditModel(field?: string) {
		const { schema } = this.props;
		if (!field) {
			field = Object.keys(schema.properties!).shift()!;
		}
		const operator = this.getOperators(field).shift()!.slug;
		return {
			field,
			operator,
			value: '',
		};
	}

	getOperators(field: string) {
		const { schema } = this.props;
		return SchemaSieve.getOperators(schema, field);
	}

	setEditField(field: string, index: number) {
		const currentEdit = this.state.edit.slice();
		currentEdit.splice(index, 1, this.getCleanEditModel(field));
		this.setState({
			edit: currentEdit,
		});
	}

	setEditOperator(operator: string, index: number) {
		const currentEdit = this.state.edit.slice();
		const item = currentEdit[index];
		currentEdit.splice(index, 1, { ...item, operator });
		this.setState({
			edit: currentEdit,
		});
	}

	setEditValue(value: string, index: number) {
		const currentEdit = this.state.edit.slice();
		const item = currentEdit[index];
		currentEdit.splice(index, 1, { ...item, value });
		this.setState({
			edit: currentEdit,
		});
	}

	addFilter() {
		const $id = this.state.editingFilter;
		const { schema } = this.props;
		const filter = SchemaSieve.createFilter(schema, this.state.edit);
		const currentFilters = this.state.filters;

		let filters: JSONSchema6[];

		if (!!$id) {
			const matchIndex = findIndex(currentFilters, { $id });
			currentFilters.splice(matchIndex, 1, filter);
			filters = currentFilters.slice();
		} else {
			filters = currentFilters.concat(filter);
		}

		this.setState(
			{
				filters,
				edit: [this.getCleanEditModel()],
				showModal: false,
				editingFilter: null,
			},
			() =>
				this.props.onFiltersUpdate &&
				this.props.onFiltersUpdate(this.state.filters),
		);
	}

	editFilter(filter: JSONSchema6) {
		const { schema } = this.props;

		const signatures = SchemaSieve.decodeFilter(schema, filter);

		this.setState({
			edit: signatures as EditModel[],
			editingFilter: filter.$id!,
			showModal: true,
		});
	}

	removeFilter({ $id }: JSONSchema6) {
		this.setState(
			prevState => ({
				filters: reject(prevState.filters, { $id }),
			}),
			() =>
				this.props.onFiltersUpdate &&
				this.props.onFiltersUpdate(this.state.filters),
		);
	}

	setFilters(filters: JSONSchema6[]) {
		this.setState(
			{ filters },
			() =>
				this.props.onFiltersUpdate &&
				this.props.onFiltersUpdate(this.state.filters),
		);
	}

	addCompound() {
		this.setState(prevState => ({
			edit: prevState.edit.concat(this.getCleanEditModel()),
		}));
	}

	removeCompound(index: number) {
		const edit = this.state.edit.slice();
		edit.splice(index, 1);
		this.setState({
			edit,
		});
	}

	saveView(name: string, scope: string | null) {
		const view: FiltersView = {
			id: utils.randomString(),
			name,
			scope,
			filters: cloneDeep(this.state.filters),
		};

		this.setState(
			prevState => ({
				views: prevState.views.concat(view),
			}),
			() =>
				this.props.onViewsUpdate && this.props.onViewsUpdate(this.state.views),
		);
	}

	deleteView({ id }: FiltersView) {
		this.setState(
			prevState => ({
				views: reject(prevState.views, { id }),
			}),
			() =>
				this.props.onViewsUpdate && this.props.onViewsUpdate(this.state.views),
		);
	}

	setSimpleSearch(term: string) {
		this.setState(prevState => {
			const newFilters = term
				? SchemaSieve.upsertFullTextSearch(
						this.props.schema,
						prevState.filters,
						term,
					)
				: SchemaSieve.removeFullTextSearch(prevState.filters);

			return {
				searchString: term,
				filters: newFilters,
			};
		}, () => this.props.onFiltersUpdate && this.props.onFiltersUpdate(this.state.filters));
	}

	render() {
		const { filters } = this.state;

		return (
			<FilterWrapper mb={3}>
				<Flex justify="space-between">
					<Button
						disabled={this.props.disabled}
						primary
						onClick={() =>
							this.setState({ showModal: true, editingFilter: null })
						}
						{...this.props.addFilterButtonProps}
					>
						<FaFilter style={{ marginRight: 10 }} />
						Add filter
					</Button>

					<SimpleSearchBox>
						<input
							style={{ color: this.props.dark ? '#fff' : undefined }}
							disabled={this.props.disabled}
							placeholder="Search entries..."
							value={this.state.searchString}
							onChange={e => this.setSimpleSearch(e.target.value)}
						/>
						<FaSearch className="search-icon" name="search" />
					</SimpleSearchBox>

					<ViewsMenu
						buttonProps={this.props.viewsMenuButtonProps}
						disabled={this.props.disabled}
						views={this.state.views || []}
						schema={this.props.schema}
						setFilters={filters => this.setFilters(filters)}
						deleteView={view => this.deleteView(view)}
					/>
				</Flex>

				{this.state.showModal && (
					<div>
						<Modal
							title="Filter by"
							cancel={() => this.setState({ showModal: false })}
							done={() => this.addFilter()}
							action="Save"
						>
							{map(this.state.edit, ({ field, operator, value }, index) => {
								const operators = this.getOperators(field);

								return (
									<RelativeBox key={index}>
										{index > 0 && <Txt my={2}>OR</Txt>}
										<Flex>
											<Select
												value={field}
												onChange={(v: any) =>
													this.setEditField(v.target.value, index)
												}
											>
												{map(
													this.props.schema.properties,
													(s: JSONSchema6, field) => (
														<option value={field}>{s.title || field}</option>
													),
												)}
											</Select>

											{operators.length === 1 && (
												<Txt mx={1} p="7px 20px 0">
													{operators[0].label}
												</Txt>
											)}

											{operators.length > 1 && (
												<Select
													ml={1}
													value={operator}
													onChange={(v: any) =>
														this.setEditOperator(v.target.value, index)
													}
												>
													{map(operators, ({ slug, label }) => (
														<option value={slug}>{label}</option>
													))}
												</Select>
											)}

											<FilterInput
												operator={operator}
												value={value}
												schema={
													this.props.schema.properties![field] as JSONSchema6
												}
												onUpdate={(v: any) => this.setEditValue(v, index)}
											/>
										</Flex>
										{index > 0 && (
											<ExtraRuleDeleteBtn
												onClick={() => this.removeCompound(index)}
											/>
										)}
									</RelativeBox>
								);
							})}
							<Button
								mb={2}
								mt={4}
								primary
								underline
								onClick={() => this.addCompound()}
							>
								Add alternative
							</Button>
						</Modal>
					</div>
				)}

				{!!filters.length &&
					!this.props.disabled && (
						<Summary
							edit={(filter: JSONSchema6) => this.editFilter(filter)}
							delete={(filter: JSONSchema6) => this.removeFilter(filter)}
							saveView={(name, scope) => this.saveView(name, scope)}
							filters={filters}
							views={this.state.views || []}
							schema={this.props.schema}
							dark={!!this.props.dark}
						/>
					)}
			</FilterWrapper>
		);
	}
}

export default Filters;
