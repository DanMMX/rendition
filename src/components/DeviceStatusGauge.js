import * as React from 'react'
const filter = require('lodash/filter')
import { withTheme } from 'styled-components'
import { compose } from 'recompose'
import * as ResinDeviceStatus from 'resin-device-status'
import Gauge from './Gauge'

import asRendition from '../asRendition'
import theme from '../theme'

const { colors } = theme

const pathFillColor = status => {
  if (status === 'idle') {
    return colors.statusIdle.main
  }

  if (status === 'configuring') {
    return colors.statusConfiguring.main
  }

  if (status === 'updating') {
    return colors.statusUpdating.main
  }

  if (status === 'post-provisioning') {
    return colors.statusPostProvisioning.main
  }

  if (status === 'offline') {
    return colors.statusOffline.main
  }
}

const DeviceStatusGauge = ({ devices, ...props }) => {
  const data = []
  const allDevices = devices || []

  ResinDeviceStatus.statuses.forEach(status => {
    const devices = filter(
      allDevices,
      d => ResinDeviceStatus.getStatus(d).key === status.key
    )

    data.push({
      key: status.key,
      name: status.name,
      value: devices.length,
      color: pathFillColor(status.key)
    })
  })

  return (
    <Gauge
      data={data}
      title='Total Devices'
      placeholderColor={colors.statusOffline.main}
    />
  )
}

export default compose(withTheme, asRendition)(DeviceStatusGauge)
