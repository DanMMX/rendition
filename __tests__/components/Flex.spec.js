import React from 'react';
import renderer from 'react-test-renderer';
import Provider from '../../src/components/Provider'
import { Flex } from '../../src/components/Grid'

test('Flex renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Flex>Flex</Flex>
    </Provider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
