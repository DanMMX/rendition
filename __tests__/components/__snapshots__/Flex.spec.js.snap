// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <div
    className="sc-gZMcBi cmZZgb sc-dnqmqq eheJhs sc-htpNat jNWwaj"
  >
    Flex
  </div>
</div>
`;
